scripts/cluster_network.py

```diff
@@ -173,6 +173,50 @@ def weighting_for_country(n, x):
     return (w * (100. / w.max())).clip(lower=1.).astype(int)
 
 
+def get_feature_for_hac(n, buses_i=None, feature=None):
+
+    if buses_i is None:
+        buses_i = n.buses.index
+
+    if feature is None:
+        feature = "solar+onwind-time"
+
+    carriers = feature.split('-')[0].split('+')
+    if "offwind" in carriers:
+        carriers.remove("offwind")
+        carriers = np.append(carriers, network.generators.carrier.filter(like='offwind').unique())
+
+    if feature.split('-')[1] == 'cap':
+        feature_data = pd.DataFrame(index=buses_i, columns=carriers)
+        for carrier in carriers:
+            try:
+                # without simpl wildcard (bus names are "X X"):
+                feature_data[carrier] = (n.generators_t.p_max_pu.filter(like=carrier).mean()
+                                         .rename(index=lambda x: x.split(' ')[0]))
+            except:
+                # with simpl wildcard (bus names are "X X X"):
+                feature_data[carrier] = (n.generators_t.p_max_pu.filter(like=carrier).mean()
+                                         .rename(index=lambda x: x.split(' ')[0] + ' ' + x.split(' ')[1]))
```

Comment:
I by off-chance had a small look into the code, out of interest for the hierarchical agglomerative clustering. I don't fully understand the feature_data concept yet, but it looks like you are trying to get a dataframe that has buses as the index and either carriers or time (as hours) on the columns.

Can't you just use the actual information of which bus the generators are attached to from the generators df, ie. something like
```python
        feature_data = {}
        for carrier in carriers:
            gen_i = n.generators.index[n.generators.carrier == carrier]
            feature_data[carrier] = n.generators.p_max_pu[gen_i].mean().set_axis(n.generators.loc[gen_i, "bus"])
        feature_data = pd.DataFrame(feature_data)
```
and
```python
        feature_data = []
        p_max_pu = n.generators_t.p_max_pu.set_axis(n.generators_t.p_max_pu.index.astype(str))
        for carrier in carriers:
            gen_i = n.generators.index[n.generators.carrier == carrier]
            feature_data.append(p_max_pu[gen_i].set_axis(n.generators.loc[gen_i, "bus"], axis="columns"))
        feature_data = pd.concat(feature_data).T
```

Reply:
Not sure how this behaves with `windoff-ac/dc`. That might be different from your implementation.

Reply:
There are actually a few style things I don't like yet in this pr, and it's lying around for too long. Therefore: Thanks for your suggestion, will try that! Getting a df with the buses as idx and carriers or time (in hours) on the cols is correct.

Reply:
> Not sure how this behaves with `windoff-ac/dc`. That might be different from your implementation.

could resolve this with a `df.query.`